File: src/performance/performance-profiler.ts
/**
 * Performance Profiler
 * 시스템 성능 분석 및 병목점 식별
 */

import { EventEmitter } from 'eventemitter3';

export interface PerformanceMetric {
  name: string;
  startTime: number;
  endTime?: number;
  duration?: number;
  metadata?: Record<string, any>;
}

export interface MemorySnapshot {
  timestamp: number;
  heapUsed: number;
  heapTotal: number;
  external: number;
  rss: number;
  arrayBuffers: number;
}

export interface CPUSnapshot {
  timestamp: number;
  userTime: number;
  systemTime: number;
  cpuUsage: number;
}

export interface ProfilerStats {
  averageResponseTime: number;
  maxResponseTime: number;
  minResponseTime: number;
  memoryTrend: 'increasing' | 'stable' | 'decreasing';
  memoryLeakPotential: number; // 0-100 score
  bottlenecks: BottleneckInfo[];
  recommendations: string[];
}

export interface BottleneckInfo {
  operation: string;
  averageDuration: number;
  callCount: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
  impact: string;
}

export class PerformanceProfiler extends EventEmitter {
  private metrics: Map<string, PerformanceMetric> = new Map();
  private completedMetrics: PerformanceMetric[] = [];
  private memorySnapshots: MemorySnapshot[] = [];
  private cpuSnapshots: CPUSnapshot[] = [];
  private isMonitoring = false;
  private monitoringInterval?: NodeJS.Timeout;
  private maxHistorySize = 1000;

  constructor() {
    super();
    this.setupMemoryMonitoring();
  }

  /**
   * 성능 메트릭 추적 시작
   */
  startMetric(name: string, metadata?: Record<string, any>): string {
    const metricId = `${name}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    
    const metric: PerformanceMetric = {
      name,
      startTime: performance.now(),
      metadata
    };

    this.metrics.set(metricId, metric);
    return metricId;
  }

  /**
   * 성능 메트릭 추적 종료
   */
  endMetric(metricId: string): PerformanceMetric | null {
    const metric = this.metrics.get(metricId);
    if (!metric) {
      return null;
    }

    const endTime = performance.now();
    metric.endTime = endTime;
    metric.duration = endTime - metric.startTime;

    this.metrics.delete(metricId);
    this.completedMetrics.push(metric);

    // 히스토리 크기 제한
    if (this.completedMetrics.length > this.maxHistorySize) {
      this.completedMetrics = this.completedMetrics.slice(-this.maxHistorySize);
    }

    // 성능 이벤트 발행
    this.emit('metricCompleted', metric);

    // 병목 현상 감지
    if (metric.duration! > 1000) { // 1초 이상
      this.emit('bottleneckDetected', {
        operation: metric.name,
        duration: metric.duration,
        severity: this.getBottleneckSeverity(metric.duration!)
      });
    }

    return metric;
  }

  /**
   * 함수 실행 시간 측정 데코레이터
   */
  measureAsync<T>(name: string, fn: () => Promise<T>): Promise<T> {
    return new Promise(async (resolve, reject) => {
      const metricId = this.startMetric(name);
      
      try {
        const result = await fn();
        this.endMetric(metricId);
        resolve(result);
      } catch (error) {
        this.endMetric(metricId);
        reject(error);
      }
    });
  }

  /**
   * 동기 함수 실행 시간 측정
   */
  measureSync<T>(name: string, fn: () => T): T {
    const metricId = this.startMetric(name);
    
    try {
      const result = fn();
      this.endMetric(metricId);
      return result;
    } catch (error) {
      this.endMetric(metricId);
      throw error;
    }
  }

  /**
   * 시스템 모니터링 시작
   */
  startMonitoring(intervalMs = 5000): void {
    if (this.isMonitoring) {
      return;
    }

    this.isMonitoring = true;
    this.monitoringInterval = setInterval(() => {
      this.captureMemorySnapshot();
      this.captureCPUSnapshot();
    }, intervalMs);

    this.emit('monitoringStarted');
  }

  /**
   * 시스템 모니터링 중지
   */
  stopMonitoring(): void {
    if (!this.isMonitoring) {
      return;
    }

    this.isMonitoring = false;
    if (this.monitoringInterval) {
      clearInterval(this.monitoringInterval);
      this.monitoringInterval = undefined;
    }

    this.emit('monitoringStopped');
  }

  /**
   * 메모리 스냅샷 캡처
   */
  private captureMemorySnapshot(): void {
    const memUsage = process.memoryUsage();
    
    const snapshot: MemorySnapshot = {
      timestamp: Date.now(),
      heapUsed: memUsage.heapUsed,
      heapTotal: memUsage.heapTotal,
      external: memUsage.external,
      rss: memUsage.rss,
      arrayBuffers: memUsage.arrayBuffers
    };

    this.memorySnapshots.push(snapshot);

    // 히스토리 크기 제한
    if (this.memorySnapshots.length > this.maxHistorySize) {
      this.memorySnapshots = this.memorySnapshots.slice(-this.maxHistorySize);
    }

    // 메모리 누수 경고
    if (this.detectMemoryLeak()) {
      this.emit('memoryLeakDetected', {
        currentHeapUsed: snapshot.heapUsed,
        trend: this.getMemoryTrend(),
        severity: this.getMemoryLeakSeverity()
      });
    }
  }

  /**
   * CPU 스냅샷 캡처
   */
  private captureCPUSnapshot(): void {
    const cpuUsage = process.cpuUsage();
    
    const snapshot: CPUSnapshot = {
      timestamp: Date.now(),
      userTime: cpuUsage.user,
      systemTime: cpuUsage.system,
      cpuUsage: (cpuUsage.user + cpuUsage.system) / 1000000 // 마이크로초를 초로 변환
    };

    this.cpuSnapshots.push(snapshot);

    // 히스토리 크기 제한
    if (this.cpuSnapshots.length > this.maxHistorySize) {
      this.cpuSnapshots = this.cpuSnapshots.slice(-this.maxHistorySize);
    }
  }

  /**
   * 성능 통계 생성
   */
  getStats(): ProfilerStats {
    const durations = this.completedMetrics
      .filter(m => m.duration !== undefined)
      .map(m => m.duration!);

    const bottlenecks = this.analyzeBottlenecks();
    
    return {
      averageResponseTime: durations.length > 0 ? durations.reduce((a, b) => a + b, 0) / durations.length : 0,
      maxResponseTime: durations.length > 0 ? Math.max(...durations) : 0,
      minResponseTime: durations.length > 0 ? Math.min(...durations) : 0,
      memoryTrend: this.getMemoryTrend(),
      memoryLeakPotential: this.getMemoryLeakPotential(),
      bottlenecks,
      recommendations: this.generateRecommendations(bottlenecks)
    };
  }

  /**
   * 병목 현상 분석
   */
  private analyzeBottlenecks(): BottleneckInfo[] {
    const operationStats = new Map<string, { durations: number[], count: number }>();

    // 작업별 통계 수집
    this.completedMetrics.forEach(metric => {
      if (!metric.duration) return;

      if (!operationStats.has(metric.name)) {
        operationStats.set(metric.name, { durations: [], count: 0 });
      }

      const stats = operationStats.get(metric.name)!;
      stats.durations.push(metric.duration);
      stats.count++;
    });

    // 병목 현상 분석
    const bottlenecks: BottleneckInfo[] = [];

    operationStats.forEach((stats, operation) => {
      const averageDuration = stats.durations.reduce((a, b) => a + b, 0) / stats.durations.length;
      
      if (averageDuration > 100) { // 100ms 이상
        bottlenecks.push({
          operation,
          averageDuration,
          callCount: stats.count,
          severity: this.getBottleneckSeverity(averageDuration),
          impact: this.getBottleneckImpact(averageDuration, stats.count)
        });
      }
    });

    return bottlenecks.sort((a, b) => b.averageDuration - a.averageDuration);
  }

  /**
   * 병목 현상 심각도 판단
   */
  private getBottleneckSeverity(duration: number): 'low' | 'medium' | 'high' | 'critical' {
    if (duration > 5000) return 'critical'; // 5초 이상
    if (duration > 2000) return 'high';     // 2초 이상
    if (duration > 500) return 'medium';    // 500ms 이상
    return 'low';
  }

  /**
   * 병목 현상 영향도 분석
   */
  private getBottleneckImpact(duration: number, callCount: number): string {
    const totalImpact = duration * callCount;
    
    if (totalImpact > 60000) { // 1분 이상의 총 영향
      return `High impact: ${(totalImpact / 1000).toFixed(1)}s total delay across ${callCount} calls`;
    } else if (totalImpact > 10000) { // 10초 이상
      return `Medium impact: ${(totalImpact / 1000).toFixed(1)}s total delay across ${callCount} calls`;
    } else {
      return `Low impact: ${(totalImpact / 1000).toFixed(1)}s total delay across ${callCount} calls`;
    }
  }

  /**
   * 메모리 추세 분석
   */
  private getMemoryTrend(): 'increasing' | 'stable' | 'decreasing' {
    if (this.memorySnapshots.length < 10) {
      return 'stable';
    }

    const recent = this.memorySnapshots.slice(-10);
    const first = recent[0].heapUsed;
    const last = recent[recent.length - 1].heapUsed;
    
    const changePercent = ((last - first) / first) * 100;
    
    if (changePercent > 10) return 'increasing';
    if (changePercent < -10) return 'decreasing';
    return 'stable';
  }

  /**
   * 메모리 누수 감지
   */
  private detectMemoryLeak(): boolean {
    if (this.memorySnapshots.length < 20) {
      return false;
    }

    const recent = this.memorySnapshots.slice(-20);
    const increases = recent.reduce((count, snapshot, index) => {
      if (index === 0) return count;
      return snapshot.heapUsed > recent[index - 1].heapUsed ? count + 1 : count;
    }, 0);

    // 20개 중 15개 이상이 증가 추세면 메모리 누수 의심
    return increases >= 15;
  }

  /**
   * 메모리 누수 잠재성 점수
   */
  private getMemoryLeakPotential(): number {
    if (this.memorySnapshots.length < 10) {
      return 0;
    }

    const trend = this.getMemoryTrend();
    const recent = this.memorySnapshots.slice(-10);
    const growthRate = this.calculateMemoryGrowthRate(recent);

    let score = 0;
    
    if (trend === 'increasing') score += 40;
    if (growthRate > 0.1) score += 30; // 10% 이상 증가
    if (this.detectMemoryLeak()) score += 30;

    return Math.min(score, 100);
  }

  /**
   * 메모리 증가율 계산
   */
  private calculateMemoryGrowthRate(snapshots: MemorySnapshot[]): number {
    if (snapshots.length < 2) return 0;

    const first = snapshots[0].heapUsed;
    const last = snapshots[snapshots.length - 1].heapUsed;
    
    return (last - first) / first;
  }

  /**
   * 성능 개선 권장사항 생성
   */
  private generateRecommendations(bottlenecks: BottleneckInfo[]): string[] {
    const recommendations: string[] = [];

    // 병목 현상 기반 권장사항
    bottlenecks.forEach(bottleneck => {
      switch (bottleneck.severity) {
        case 'critical':
          recommendations.push(`CRITICAL: ${bottleneck.operation} 작업 최적화 필요 (평균 ${bottleneck.averageDuration.toFixed(0)}ms)`);
          break;
        case 'high':
          recommendations.push(`HIGH: ${bottleneck.operation} 작업에 캐싱 또는 비동기 처리 적용 검토`);
          break;
        case 'medium':
          recommendations.push(`MEDIUM: ${bottleneck.operation} 작업 성능 모니터링 지속`);
          break;
      }
    });

    // 메모리 기반 권장사항
    const memoryTrend = this.getMemoryTrend();
    const memoryLeakPotential = this.getMemoryLeakPotential();

    if (memoryTrend === 'increasing' && memoryLeakPotential > 70) {
      recommendations.push('CRITICAL: 메모리 누수 가능성 높음 - 즉시 조사 필요');
    } else if (memoryTrend === 'increasing') {
      recommendations.push('WARNING: 메모리 사용량 증가 추세 - 정기 모니터링 필요');
    }

    // 일반 권장사항
    if (recommendations.length === 0) {
      recommendations.push('성능 지표가 양호합니다. 현재 상태를 유지하세요.');
    }

    return recommendations;
  }

  /**
   * 메모리 모니터링 설정
   */
  private setupMemoryMonitoring(): void {
    // 초기 메모리 스냅샷
    this.captureMemorySnapshot();

    // 메모리 압박 상황 감지
    process.on('warning', (warning) => {
      if (warning.name === 'MaxListenersExceededWarning') {
        this.emit('performanceWarning', {
          type: 'memory',
          message: 'EventEmitter 리스너 수 초과',
          severity: 'medium'
        });
      }
    });
  }

  /**
   * 리소스 정리
   */
  cleanup(): void {
    this.stopMonitoring();
    this.metrics.clear();
    this.completedMetrics.length = 0;
    this.memorySnapshots.length = 0;
    this.cpuSnapshots.length = 0;
    this.removeAllListeners();
  }

  /**
   * 성능 리포트 생성
   */
  generateReport(): {
    summary: ProfilerStats;
    details: {
      recentMetrics: PerformanceMetric[];
      memoryHistory: MemorySnapshot[];
      cpuHistory: CPUSnapshot[];
    };
  } {
    return {
      summary: this.getStats(),
      details: {
        recentMetrics: this.completedMetrics.slice(-50),
        memoryHistory: this.memorySnapshots.slice(-50),
        cpuHistory: this.cpuSnapshots.slice(-50)
      }
    };
  }
}

// 싱글톤 인스턴스
export const performanceProfiler = new PerformanceProfiler();

File: python/src/devflow_monitor/performance/performance_profiler.py
"""
Performance Profiler
시스템 성능 분석 및 병목점 식별

This module provides performance profiling functionality including
metric tracking, CPU/memory monitoring, bottleneck detection, and
latency percentile analysis.
"""

import asyncio
import gc
import statistics
import sys
import threading
import time
import uuid
from typing import Any, Callable, Optional, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class PerformanceMetric(BaseModel):
    """Performance metric data."""

    name: str
    start_time: float
    end_time: Optional[float] = None
    duration: Optional[float] = None
    metadata: Optional[dict[str, Any]] = None


class MemorySnapshot(BaseModel):
    """Memory usage snapshot."""

    timestamp: int
    heap_used: int
    heap_total: int
    external: int
    rss: int
    array_buffers: int


class CPUSnapshot(BaseModel):
    """CPU usage snapshot."""

    timestamp: int
    user_time: float
    system_time: float
    cpu_usage: float


class BottleneckInfo(BaseModel):
    """Bottleneck information."""

    operation: str
    average_duration: float
    call_count: int
    severity: str  # low, medium, high, critical
    impact: str


class ProfilerStats(BaseModel):
    """Profiler statistics."""

    average_response_time: float
    max_response_time: float
    min_response_time: float
    p50_response_time: float
    p95_response_time: float
    p99_response_time: float
    memory_trend: str  # increasing, stable, decreasing
    memory_leak_potential: int  # 0-100 score
    memory_usage: MemorySnapshot
    bottlenecks: list[BottleneckInfo]
    recommendations: list[str]


class PerformanceProfiler:
    """
    Performance Profiler for system performance analysis.

    Provides functionality for:
    - Metric tracking with start/end timing
    - CPU and memory monitoring
    - Bottleneck detection and severity classification
    - Memory leak detection
    - Latency percentile calculation (P50, P95, P99)
    - Performance recommendations generation

    Example:
        >>> profiler = PerformanceProfiler()
        >>> profiler.start_monitoring()
        >>> metric_id = profiler.start_metric("my_operation")
        >>> # ... do work ...
        >>> profiler.end_metric(metric_id)
        >>> stats = profiler.get_stats()
    """

    def __init__(self) -> None:
        """Initialize the performance profiler."""
        self._metrics: dict[str, PerformanceMetric] = {}
        self._completed_metrics: list[PerformanceMetric] = []
        self._memory_snapshots: list[MemorySnapshot] = []
        self._cpu_snapshots: list[CPUSnapshot] = []
        self._is_monitoring = False
        self._monitoring_task: Optional[asyncio.Task[None]] = None
        self._monitoring_thread: Optional[threading.Thread] = None
        self._max_history_size = 1000
        self._performance_history: list[ProfilerStats] = []
        self._lock = threading.Lock()
        self._event_handlers: list[tuple[str, Any]] = []
        self._stop_monitoring_flag = False

        # Capture initial memory snapshot
        self._capture_memory_snapshot()

    def on_event(self, event_name: str, handler: Any) -> None:
        """Register an event handler."""
        self._event_handlers.append((event_name, handler))

    def _emit_event(self, event_name: str, data: Any) -> None:
        """Emit an event to registered handlers."""
        for name, handler in self._event_handlers:
            if name == event_name:
                try:
                    handler(data)
                except Exception:
                    pass

    def start_metric(
        self, name: str, metadata: Optional[dict[str, Any]] = None
    ) -> str:
        """
        Start tracking a performance metric.

        Args:
            name: Name of the operation being measured
            metadata: Optional metadata for the metric

        Returns:
            Unique metric ID for ending the metric
        """
        metric_id = f"{name}_{int(time.time() * 1000)}_{uuid.uuid4().hex[:9]}"

        metric = PerformanceMetric(
            name=name,
            start_time=time.perf_counter() * 1000,  # Convert to ms
            metadata=metadata,
        )

        with self._lock:
            self._metrics[metric_id] = metric

        return metric_id

    def end_metric(self, metric_id: str) -> Optional[PerformanceMetric]:
        """
        End tracking a performance metric.

        Args:
            metric_id: The metric ID returned from start_metric

        Returns:
            The completed metric or None if not found
        """
        with self._lock:
            metric = self._metrics.get(metric_id)
            if not metric:
                return None

            end_time = time.perf_counter() * 1000  # Convert to ms
            metric.end_time = end_time
            metric.duration = end_time - metric.start_time

            del self._metrics[metric_id]
            self._completed_metrics.append(metric)

            # Limit history size
            if len(self._completed_metrics) > self._max_history_size:
                self._completed_metrics = self._completed_metrics[
                    -self._max_history_size :
                ]

        # Emit metric completed event
        self._emit_event("metric_completed", metric.model_dump())

        # Detect bottleneck (> 1000ms)
        if metric.duration and metric.duration > 1000:
            self._emit_event(
                "bottleneck_detected",
                {
                    "operation": metric.name,
                    "duration": metric.duration,
                    "severity": self._get_bottleneck_severity(metric.duration),
                },
            )

        return metric

    async def measure_async(
        self, name: str, fn: Callable[[], Any]
    ) -> Any:
        """
        Measure execution time of an async function.

        Args:
            name: Name for the metric
            fn: Async function to measure

        Returns:
            Result of the function
        """
        metric_id = self.start_metric(name)
        try:
            result = await fn()
            self.end_metric(metric_id)
            return result
        except Exception as e:
            self.end_metric(metric_id)
            raise e

    def measure_sync(self, name: str, fn: Callable[[], T]) -> T:
        """
        Measure execution time of a sync function.

        Args:
            name: Name for the metric
            fn: Function to measure

        Returns:
            Result of the function
        """
        metric_id = self.start_metric(name)
        try:
            result = fn()
            self.end_metric(metric_id)
            return result
        except Exception as e:
            self.end_metric(metric_id)
            raise e

    def start_monitoring(self, interval_ms: int = 5000) -> None:
        """
        Start system monitoring.

        Args:
            interval_ms: Monitoring interval in milliseconds
        """
        if self._is_monitoring:
            return

        self._is_monitoring = True
        self._stop_monitoring_flag = False

        def monitor_loop() -> None:
            while not self._stop_monitoring_flag:
                self._capture_memory_snapshot()
                self._capture_cpu_snapshot()
                time.sleep(interval_ms / 1000)

        self._monitoring_thread = threading.Thread(
            target=monitor_loop, daemon=True
        )
        self._monitoring_thread.start()

        self._emit_event("monitoring_started", {"interval_ms": interval_ms})

    def stop_monitoring(self) -> None:
        """Stop system monitoring."""
        if not self._is_monitoring:
            return

        self._is_monitoring = False
        self._stop_monitoring_flag = True

        if self._monitoring_thread:
            self._monitoring_thread.join(timeout=1)
            self._monitoring_thread = None

        self._emit_event("monitoring_stopped", {})

    def _capture_memory_snapshot(self) -> None:
        """Capture a memory usage snapshot."""
        try:
            import psutil

            process = psutil.Process()
            mem_info = process.memory_info()

            snapshot = MemorySnapshot(
                timestamp=int(time.time() * 1000),
                heap_used=mem_info.rss,
                heap_total=mem_info.vms,
                external=0,  # Not directly available in Python
                rss=mem_info.rss,
                array_buffers=0,  # Not applicable in Python
            )
        except ImportError:
            # Fallback without psutil
            snapshot = MemorySnapshot(
                timestamp=int(time.time() * 1000),
                heap_used=sys.getsizeof({}),  # Minimal estimate
                heap_total=0,
                external=0,
                rss=0,
                array_buffers=0,
            )

        with self._lock:
            self._memory_snapshots.append(snapshot)

            # Limit history size
            if len(self._memory_snapshots) > self._max_history_size:
                self._memory_snapshots = self._memory_snapshots[
                    -self._max_history_size :
                ]

        # Detect memory leak
        if self._detect_memory_leak():
            self._emit_event(
                "memory_leak_detected",
                {
                    "current_heap_used": snapshot.heap_used,
                    "trend": self._get_memory_trend(),
                    "severity": self._get_memory_leak_severity(),
                },
            )

    def _capture_cpu_snapshot(self) -> None:
        """Capture a CPU usage snapshot."""
        try:
            import psutil

            cpu_times = psutil.Process().cpu_times()
            cpu_percent = psutil.Process().cpu_percent()

            snapshot = CPUSnapshot(
                timestamp=int(time.time() * 1000),
                user_time=cpu_times.user,
                system_time=cpu_times.system,
                cpu_usage=cpu_percent,
            )
        except ImportError:
            snapshot = CPUSnapshot(
                timestamp=int(time.time() * 1000),
                user_time=0,
                system_time=0,
                cpu_usage=0,
            )

        with self._lock:
            self._cpu_snapshots.append(snapshot)

            # Limit history size
            if len(self._cpu_snapshots) > self._max_history_size:
                self._cpu_snapshots = self._cpu_snapshots[
                    -self._max_history_size :
                ]

    def get_stats(self) -> ProfilerStats:
        """
        Get profiler statistics.

        Returns:
            Complete profiler statistics including bottlenecks and recommendations
        """
        with self._lock:
            durations = [
                m.duration
                for m in self._completed_metrics
                if m.duration is not None
            ]

        bottlenecks = self._analyze_bottlenecks()

        # Get latest memory snapshot or create default
        if self._memory_snapshots:
            latest_snapshot = self._memory_snapshots[-1]
        else:
            latest_snapshot = MemorySnapshot(
                timestamp=int(time.time() * 1000),
                heap_used=0,
                heap_total=0,
                external=0,
                rss=0,
                array_buffers=0,
            )

        # Calculate percentiles
        if durations:
            sorted_durations = sorted(durations)
            p50 = self._percentile(sorted_durations, 50)
            p95 = self._percentile(sorted_durations, 95)
            p99 = self._percentile(sorted_durations, 99)
            avg = statistics.mean(durations)
            max_time = max(durations)
            min_time = min(durations)
        else:
            p50 = p95 = p99 = avg = max_time = min_time = 0

        return ProfilerStats(
            average_response_time=avg,
            max_response_time=max_time,
            min_response_time=min_time,
            p50_response_time=p50,
            p95_response_time=p95,
            p99_response_time=p99,
            memory_trend=self._get_memory_trend(),
            memory_leak_potential=self._get_memory_leak_potential(),
            memory_usage=latest_snapshot,
            bottlenecks=bottlenecks,
            recommendations=self._generate_recommendations(bottlenecks),
        )

    def _percentile(self, sorted_data: list[float], percentile: int) -> float:
        """Calculate percentile from sorted data."""
        if not sorted_data:
            return 0
        index = (len(sorted_data) - 1) * percentile / 100
        lower = int(index)
        upper = lower + 1
        if upper >= len(sorted_data):
            return sorted_data[-1]
        weight = index - lower
        return sorted_data[lower] * (1 - weight) + sorted_data[upper] * weight

    def _analyze_bottlenecks(self) -> list[BottleneckInfo]:
        """Analyze operations for bottlenecks."""
        operation_stats: dict[str, dict[str, Any]] = {}

        with self._lock:
            for metric in self._completed_metrics:
                if metric.duration is None:
                    continue

                if metric.name not in operation_stats:
                    operation_stats[metric.name] = {"durations": [], "count": 0}

                stats = operation_stats[metric.name]
                stats["durations"].append(metric.duration)
                stats["count"] += 1

        bottlenecks: list[BottleneckInfo] = []

        for operation, stats in operation_stats.items():
            if not stats["durations"]:
                continue

            avg_duration = statistics.mean(stats["durations"])

            # Flag as bottleneck if average > 100ms
            if avg_duration > 100:
                bottlenecks.append(
                    BottleneckInfo(
                        operation=operation,
                        average_duration=avg_duration,
                        call_count=stats["count"],
                        severity=self._get_bottleneck_severity(avg_duration),
                        impact=self._get_bottleneck_impact(
                            avg_duration, stats["count"]
                        ),
                    )
                )

        # Sort by average duration descending
        bottlenecks.sort(key=lambda b: b.average_duration, reverse=True)
        return bottlenecks

    def _get_bottleneck_severity(
        self, duration: float
    ) -> str:
        """Determine bottleneck severity based on duration."""
        if duration > 5000:
            return "critical"
        if duration > 2000:
            return "high"
        if duration > 500:
            return "medium"
        return "low"

    def _get_bottleneck_impact(self, duration: float, call_count: int) -> str:
        """Calculate bottleneck impact description."""
        total_impact = duration * call_count

        if total_impact > 60000:
            return f"High impact: {total_impact / 1000:.1f}s total delay across {call_count} calls"
        elif total_impact > 10000:
            return f"Medium impact: {total_impact / 1000:.1f}s total delay across {call_count} calls"
        else:
            return f"Low impact: {total_impact / 1000:.1f}s total delay across {call_count} calls"

    def _get_memory_trend(self) -> str:
        """Analyze memory usage trend."""
        with self._lock:
            if len(self._memory_snapshots) < 10:
                return "stable"

            recent = self._memory_snapshots[-10:]

        first = recent[0].heap_used if recent else 0
        last = recent[-1].heap_used if recent else 0

        if first == 0:
            return "stable"

        change_percent = ((last - first) / first) * 100

        if change_percent > 10:
            return "increasing"
        if change_percent < -10:
            return "decreasing"
        return "stable"

    def _detect_memory_leak(self) -> bool:
        """Detect potential memory leak."""
        with self._lock:
            if len(self._memory_snapshots) < 20:
                return False

            recent = self._memory_snapshots[-20:]

        increases = sum(
            1
            for i in range(1, len(recent))
            if recent[i].heap_used > recent[i - 1].heap_used
        )

        # If 15+ out of 20 are increases, suspect memory leak
        return increases >= 15

    def _get_memory_leak_potential(self) -> int:
        """Calculate memory leak potential score (0-100)."""
        with self._lock:
            if len(self._memory_snapshots) < 10:
                return 0

            recent = self._memory_snapshots[-10:]

        trend = self._get_memory_trend()
        growth_rate = self._calculate_memory_growth_rate(recent)

        score = 0

        if trend == "increasing":
            score += 40
        if growth_rate > 0.1:
            score += 30
        if self._detect_memory_leak():
            score += 30

        return min(score, 100)

    def _calculate_memory_growth_rate(
        self, snapshots: list[MemorySnapshot]
    ) -> float:
        """Calculate memory growth rate."""
        if len(snapshots) < 2:
            return 0

        first = snapshots[0].heap_used
        last = snapshots[-1].heap_used

        if first == 0:
            return 0

        return (last - first) / first

    def _get_memory_leak_severity(
        self,
    ) -> str:
        """Calculate memory leak severity."""
        score = self._calculate_memory_leak_score()

        if score >= 80:
            return "critical"
        if score >= 60:
            return "high"
        if score >= 40:
            return "medium"
        if score >= 20:
            return "low"
        return "none"

    def _calculate_memory_leak_score(self) -> int:
        """Calculate memory leak score."""
        with self._lock:
            recent_metrics = self._performance_history[-10:]

        if len(recent_metrics) < 3:
            return 0

        memory_increase_count = 0
        total_memory_increase = 0

        for i in range(1, len(recent_metrics)):
            current = recent_metrics[i]
            previous = recent_metrics[i - 1]

            memory_diff = (
                current.memory_usage.heap_used - previous.memory_usage.heap_used
            )
            if memory_diff > 0:
                memory_increase_count += 1
                total_memory_increase += memory_diff

        if memory_increase_count == 0:
            return 0

        increase_ratio = memory_increase_count / (len(recent_metrics) - 1)
        avg_increase = total_memory_increase / memory_increase_count

        score = 0

        # Increase frequency (max 50 points)
        score += int(increase_ratio * 50)

        # Average increase amount (max 50 points)
        max_increase_threshold = 10 * 1024 * 1024  # 10MB
        score += min(int((avg_increase / max_increase_threshold) * 50), 50)

        return min(score, 100)

    def _generate_recommendations(
        self, bottlenecks: list[BottleneckInfo]
    ) -> list[str]:
        """Generate performance recommendations."""
        recommendations: list[str] = []

        # Bottleneck-based recommendations
        for bottleneck in bottlenecks:
            if bottleneck.severity == "critical":
                recommendations.append(
                    f"CRITICAL: {bottleneck.operation} operation needs optimization "
                    f"(avg {bottleneck.average_duration:.0f}ms)"
                )
            elif bottleneck.severity == "high":
                recommendations.append(
                    f"HIGH: Consider caching or async processing for {bottleneck.operation}"
                )
            elif bottleneck.severity == "medium":
                recommendations.append(
                    f"MEDIUM: Continue monitoring {bottleneck.operation} performance"
                )

        # Memory-based recommendations
        memory_trend = self._get_memory_trend()
        memory_leak_potential = self._get_memory_leak_potential()

        if memory_trend == "increasing" and memory_leak_potential > 70:
            recommendations.append(
                "CRITICAL: High memory leak potential - immediate investigation needed"
            )
        elif memory_trend == "increasing":
            recommendations.append(
                "WARNING: Memory usage trending up - regular monitoring advised"
            )

        # Default recommendation if all is well
        if not recommendations:
            recommendations.append(
                "Performance metrics are healthy. Maintain current state."
            )

        return recommendations

    def generate_report(self) -> dict[str, Any]:
        """
        Generate a complete performance report.

        Returns:
            Report with summary and detailed metrics
        """
        with self._lock:
            recent_metrics = self._completed_metrics[-50:]
            memory_history = self._memory_snapshots[-50:]
            cpu_history = self._cpu_snapshots[-50:]

        return {
            "summary": self.get_stats().model_dump(),
            "details": {
                "recent_metrics": [m.model_dump() for m in recent_metrics],
                "memory_history": [s.model_dump() for s in memory_history],
                "cpu_history": [s.model_dump() for s in cpu_history],
            },
        }

    def cleanup(self) -> None:
        """Clean up resources."""
        self.stop_monitoring()

        with self._lock:
            self._metrics.clear()
            self._completed_metrics.clear()
            self._memory_snapshots.clear()
            self._cpu_snapshots.clear()

        self._event_handlers.clear()


# Singleton instance
_performance_profiler: Optional[PerformanceProfiler] = None


def get_performance_profiler() -> PerformanceProfiler:
    """Get or create the performance profiler singleton."""
    global _performance_profiler
    if _performance_profiler is None:
        _performance_profiler = PerformanceProfiler()
    return _performance_profiler


# Default singleton for convenience
performance_profiler = PerformanceProfiler()
